### PR TITLE
Throttle Eth_ module queue

### LINKED ISSUE #5923: "Unable to rent an instance of IEthRpcModule. Too many concurrent requests." error in response
**Description**
There is an error response comes from node for the eth_call. It's **not resproduced always** but thrown sometimes (request is provided beflow).

Node is using Gnosis Mainnet. 

**Steps to Reproduce**
Steps to reproduce:

1. Send JSON rpc request to the node (please see the request in [nm_request_too_many_concurrent.txt](https://github.com/NethermindEth/nethermind/files/12050967/nm_request_too_many_concurrent.txt) attachment) 

**Actual behavior**
```

{
  "jsonrpc": "2.0",
  "error": {
    "code": -32017,
    "message": "Timeout",
    "data": "Nethermind.JsonRpc.Modules.ModuleRentalTimeoutException: Unable to rent an instance of IEthRpcModule. Too many concurrent requests.\n   at Nethermind.JsonRpc.Modules.BoundedModulePool`1.SlowPath() in /src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:line 45\n   at Nethermind.JsonRpc.Modules.RpcModuleProvider.<>c__DisplayClass18_0`1.<<Register>b__0>d.MoveNext() in /src/Nethermind/Nethermind.JsonRpc/Modules/RpcModuleProvider.cs:line 59\n--- End of stack trace from previous location ---\n   at Nethermind.JsonRpc.JsonRpcService.ExecuteAsync(JsonRpcRequest request, String methodName, ValueTuple`2 method, JsonRpcContext context) in /src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:line 235\n   at Nethermind.JsonRpc.JsonRpcService.ExecuteRequestAsync(JsonRpcRequest rpcRequest, JsonRpcContext context) in /src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:line 107\n   at Nethermind.JsonRpc.JsonRpcService.SendRequestAsync(JsonRpcRequest rpcRequest, JsonRpcContext context) in /src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:line 97"
  },
  "id": 12

}

```

**Expected behavior**
No error in response

**Desktop (please complete the following information):**
Operating System: Linux
Version: "nethermind": "1.19.3"
Installation Method: Docker
Consensus Client: "lighthouse": "v4.2.0"

### PR DESCRIPTION
**Pull request by MarekM25**
## Fixes #5923

With spam testing

**Before change:**
Requests per second: 100
Requests sent in total: 100000
Succeeded requests: 4818
Failed requests: **95182**

**After the change:**
Requests per second: 100
Requests sent in total: 100000
Succeeded requests: 57566
Failed requests: **42434**

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

We did load testing with eth_call and testing on gateway infra

_Optional. Remove if not applicable._

#### Requires documentation update

- [x] Yes
- [ ] No

Auto-updated with a new config parameter. However, would be great to introduce articles Nethermind for RPC Providers

#### Requires explanation in Release Notes

- [x] Yes
- [ ] No

ToDo...

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 1]
```diff
@@ -35,11 +46,27 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
         private async Task<T> SlowPath()
         {
+            bool pendingRequestLimitsEnabled = _maxPendingSharedRequests > 0;
+            if (pendingRequestLimitsEnabled && _rpcCallsPending > _maxPendingSharedRequests)
+            {
+                // ToDo Change this exception
+                throw new ModuleRentalTimeoutException($"Unable to start new pending requests for {typeof(T).Name}. Too many pending requests. Pending calls {_rpcCallsPending}.");
+            }
+
+            ++_rpcCallsPending;
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:56]: I know that to be 100% correct here we should have an extra semaphore. However, not sure if being 100% correct is worth adding an extra semaphore due to potential performance hits. However, definitely, I have to think about it more and I planned to do so

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 2]
```diff
@@ -15,9 +19,16 @@ public class BoundedModulePool<T> : IRpcModulePool<T> where T : IRpcModule
         private readonly Task<T> _sharedAsTask;
         private readonly ConcurrentQueue<T> _pool = new();
         private readonly SemaphoreSlim _semaphore;
+        private readonly ILogger _logger;
+        private int _rpcPendingCalls = 0;
+        private readonly int _maxPendingSharedRequests = 0;
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:24]:
```suggestion
        private readonly int _requestQueueLimit = 0;
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 3]
```diff
@@ -15,9 +19,16 @@ public class BoundedModulePool<T> : IRpcModulePool<T> where T : IRpcModule
         private readonly Task<T> _sharedAsTask;
         private readonly ConcurrentQueue<T> _pool = new();
         private readonly SemaphoreSlim _semaphore;
+        private readonly ILogger _logger;
+        private int _rpcPendingCalls = 0;
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:23]:
```suggestion
        private int _rpcQueuedCalls = 0;
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 4]
```diff
@@ -15,9 +19,16 @@ public class BoundedModulePool<T> : IRpcModulePool<T> where T : IRpcModule
...
+        private readonly int _maxPendingSharedRequests = 0;
+        private readonly bool _pendingRequestLimitsEnabled = false;
 
-        public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, int timeout)
+        public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, int timeout, ILogManager logManager, int maxPendingSharedRequests = 0)
         {
+            _maxPendingSharedRequests = maxPendingSharedRequests;
+            _pendingRequestLimitsEnabled = _maxPendingSharedRequests > 0;
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:30]:
```suggestion
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 5]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
 
         private async Task<T> SlowPath()
         {
+            if (_pendingRequestLimitsEnabled && _rpcPendingCalls > _maxPendingSharedRequests)
+            {
+                throw new LimitExceededException($"Unable to start new pending requests for {typeof(T).Name}. Too many pending requests. Pending calls {_rpcPendingCalls}.");
+            }
+
+            IncrementRpcPendingCalls();
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:54]:
```suggestion
            IncrementRpcQueuedCalls();
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 6]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
+            if (_logger.IsTrace)
+                _logger.Trace($"{typeof(T).Name} Pending RPC requests {_rpcPendingCalls}");
+
             if (!await _semaphore.WaitAsync(_timeout))
             {
+                DecrementRpcPendingCalls();
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:60]:
```suggestion
                DecrementRpcQueuedCalls();
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 7]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
                 throw new ModuleRentalTimeoutException($"Unable to rent an instance of {typeof(T).Name}. Too many concurrent requests.");
             }
 
+            DecrementRpcPendingCalls();
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:64]:
```suggestion
            DecrementRpcQueuedCalls();
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 8]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
             _pool.TryDequeue(out T result);
             return result;
         }
 
+        private void IncrementRpcPendingCalls()
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:69]:
```suggestion
        private void IncrementRpcQueuedCalls()
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 9]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
+        {
+            if (_pendingRequestLimitsEnabled)
+                Interlocked.Increment(ref _rpcPendingCalls);
+        }
+
+        private void DecrementRpcPendingCalls()
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:75]:
```suggestion
        private void DecrementRpcQueuedCalls()
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 10]
```diff
@@ -15,9 +19,16 @@ public class BoundedModulePool<T> : IRpcModulePool<T> where T : IRpcModule
...
         private readonly SemaphoreSlim _semaphore;
+        private readonly ILogger _logger;
+        private int _rpcPendingCalls = 0;
+        private readonly int _maxPendingSharedRequests = 0;
+        private readonly bool _pendingRequestLimitsEnabled = false;
 
-        public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, int timeout)
+        public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, int timeout, ILogManager logManager, int maxPendingSharedRequests = 0)
         {
+            _maxPendingSharedRequests = maxPendingSharedRequests;
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:29]:
```suggestion
            _requestQueueLimit = requestQueueLimit;
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 11]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
 
         private async Task<T> SlowPath()
         {
+            if (_pendingRequestLimitsEnabled && _rpcPendingCalls > _maxPendingSharedRequests)
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:49]:
```suggestion
            if (RequestLimitsEnabled && _rpcQueuedCalls > _requestQueueLimit)
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 12]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
 
+            DecrementRpcPendingCalls();
             _pool.TryDequeue(out T result);
             return result;
         }
 
+        private void IncrementRpcPendingCalls()
+        {
+            if (_pendingRequestLimitsEnabled)
+                Interlocked.Increment(ref _rpcPendingCalls);
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:72]:
```suggestion
                Interlocked.Increment(ref _rpcQueuedCalls);
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 13]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
+        }
+
+        private void DecrementRpcPendingCalls()
+        {
+            if (_pendingRequestLimitsEnabled)
+                Interlocked.Decrement(ref _rpcPendingCalls);
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:78]:
```suggestion
                Interlocked.Decrement(ref _rpcQueuedCalls);
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 14]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
             }
 
+            DecrementRpcPendingCalls();
             _pool.TryDequeue(out T result);
             return result;
         }
 
+        private void IncrementRpcPendingCalls()
+        {
+            if (_pendingRequestLimitsEnabled)
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:71]:
```suggestion
            if (RequestLimitsEnabled)
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs [hunk 15]
```diff
@@ -35,15 +46,38 @@ public BoundedModulePool(IRpcModuleFactory<T> factory, int exclusiveCapacity, in
...
+                Interlocked.Increment(ref _rpcPendingCalls);
+        }
+
+        private void DecrementRpcPendingCalls()
+        {
+            if (_pendingRequestLimitsEnabled)
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BoundedModulePool.cs:77]:
```suggestion
            if (RequestLimitsEnabled)
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/IRpcModuleProviderExtensions.cs [hunk 16]
```diff
@@ -13,19 +14,22 @@ public static class IRpcModuleProviderExtensions
...
             ModuleFactoryBase<T> factory,
             int maxCount,
-            int timeout)
+            int timeout,
+            ILogManager logManager,
+            int maxPendingSharedRequests = 0)
             where T : IRpcModule
         {
-            rpcModuleProvider.Register(new BoundedModulePool<T>(factory, maxCount, timeout));
+            rpcModuleProvider.Register(new BoundedModulePool<T>(factory, maxCount, timeout, logManager, maxPendingSharedRequests));
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/IRpcModuleProviderExtensions.cs:22]:
```suggestion
            rpcModuleProvider.Register(new BoundedModulePool<T>(factory, maxCount, timeout, logManager, requestQueueLimit));
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/IRpcModuleProviderExtensions.cs [hunk 17]
```diff
@@ -13,19 +14,22 @@ public static class IRpcModuleProviderExtensions
             this IRpcModuleProvider rpcModuleProvider,
             ModuleFactoryBase<T> factory,
             int maxCount,
-            int timeout)
+            int timeout,
+            ILogManager logManager,
+            int maxPendingSharedRequests = 0)
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/Modules/IRpcModuleProviderExtensions.cs:19]:
```suggestion
            int requestQueueLimit = 0)
```

File: src/Nethermind/Nethermind.JsonRpc/IJsonRpcConfig.cs [hunk 18]
```diff
@@ -23,6 +23,19 @@ public interface IJsonRpcConfig : IConfig
             DefaultValue = "20000")]
         int Timeout { get; set; }
 
+        [ConfigItem(
+            Description = "The limit for maximum pending requests for non sharable calls (" +
```
Comment (benaadams) [inline src/Nethermind/Nethermind.JsonRpc/IJsonRpcConfig.cs:27]:
```suggestion
            Description = "The queued request limit for calls above the max concurrency amount for (" +
```

**Review by benaadams — APPROVED** [review]

File: src/Nethermind/Nethermind.JsonRpc/IJsonRpcConfig.cs [hunk 19]
```diff
@@ -23,6 +23,19 @@ public interface IJsonRpcConfig : IConfig
...
+                          nameof(IEthRpcModule.eth_call) + ", " +
+                          nameof(IEthRpcModule.eth_estimateGas) + ", " +
+                          nameof(IEthRpcModule.eth_getLogs) + ", " +
+                          nameof(IEthRpcModule.eth_newFilter) + ", " +
+                          nameof(IEthRpcModule.eth_newBlockFilter) + ", " +
+                          nameof(IEthRpcModule.eth_newPendingTransactionFilter) + ", " +
+                          nameof(IEthRpcModule.eth_uninstallFilter) + "). " +
+                          " If value is set to 0 limit won't be applied.",
+            DefaultValue = "500")]
+        int RequestQueueLimit { get; set; }
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.JsonRpc/IJsonRpcConfig.cs:37]: @benaadams it is only for non-shared methods, shouldn't we mention it in the config name?

Comment (benaadams) [reply src/Nethermind/Nethermind.JsonRpc/IJsonRpcConfig.cs:37]: Mentions the calls its for in the description?

User wouldn't understand what "non-shared" methods are; is an implementation detail?

**Pull request merged**